### PR TITLE
fault injection: add support for setting gRPC status

### LINKED ISSUE #10323: Fault Injection: Support gRPC status code in abort fault
***Description*:**

Currently, the Envoy [fault injection filter](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter#config-http-filters-fault-injection) supports the ability to inject an ‘abort’ fault into a response stream. These faults can be [configured](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/http/fault/v2/fault.proto#envoy-api-msg-config-filter-http-fault-v2-faultabort) with an HTTP status code that will be returned to the client when the request is failed.

For gRPC services, the status of the request is returned to the caller in a combination of both an HTTP status code and a gRPC status code, both of which are carried in HTTP/2 headers (or trailers). Envoy’s fault injection filter doesn’t currently support a way of specifying the gRPC status code to return in case of failure, instead, it implicitly maps the HTTP status code to gRPC status code based on [this](https://github.com/envoyproxy/envoy/blob/c85bf324753414c205f1006724d2f5e67dec2d76/source/common/grpc/status.cc#L6). In order to simulate specific gRPC failure modes, we need to add support for configuring gRPC status code in abort fault similar to the HTTP status code.

***Implementation details***

A new ‘grpc_status’ field will be added to the envoy fault injection filter configuration. When this field is set and the conditions for fault injection are met, Envoy will fail the gRPC call. The upstream service will not be invoked and a [‘Trailers-Only’](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) gRPC response will be sent to the client. The gRPC status code in this response will be set to the configured value.

The [utility function](https://github.com/envoyproxy/envoy/blob/9be85fbc561c373875a0102dddf9baf5a86543c1/source/common/http/utility.h#L191) which is being used to send the fault response, already has the functionality to use a specific gRPC status code instead of deriving it based on an HTTP status code, but it is currently being passed null for the gRPC status code.

Changes required to achieve this functionality will be:

- [ ] Introduce a new field “grpc_status” (uint32) in envoy’s fault filter configuration inside FaultAbort [here](https://github.com/envoyproxy/envoy/blob/139f926c1517672cdcc833e44c3b64d480e6ef39/api/envoy/config/filter/http/fault/v2/fault.proto#L30)
- [ ] Read the new field and pass it to the [utility function](https://github.com/envoyproxy/envoy/blob/9be85fbc561c373875a0102dddf9baf5a86543c1/source/common/http/utility.h#L191) via [this](https://github.com/envoyproxy/envoy/blob/c5f4302325223765b660f0f366ce25bf8570e7a5/source/extensions/filters/http/fault/fault_filter.cc#L367) when set.
- [ ] Add new unit test(s) for gRPC code [here](https://github.com/envoyproxy/envoy/blob/a0d952048c536c2da29be79df480a046ccf2be89/test/extensions/filters/http/fault/fault_filter_test.cc#L225)
- [ ] Documentation update [here](https://www.envoyproxy.io/docs/envoy/latest/api-v2/config/filter/http/fault/v2/fault.proto#config-filter-http-fault-v2-faultabort) and [here](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/fault_filter#runtime)

### PR DESCRIPTION
**Pull request by vijayrajput1**
Description: Adds an option to specify gRPC status via config and header, to be used in fault injection response. More details #10323 

Risk Level: Low (new feature)
Testing: Added unit and integration tests
Docs Changes: Updated
Release Notes: Updated
Fixes #10323

### CODE REVIEW

File: api/envoy/config/filter/http/fault/v2/fault.proto [hunk 1]
```diff
@@ -22,6 +22,7 @@ option (udpa.annotations.file_status).package_version_status = FROZEN;
 // Fault Injection :ref:`configuration overview <config_http_filters_fault_injection>`.
 // [#extension: envoy.filters.http.fault]
 
+// [#next-free-field: 6]
```
Comment (htuch) [inline api/envoy/config/filter/http/fault/v2/fault.proto:25]: v2 APIs are frozen, so this should be v3 only.

File: api/envoy/extensions/filters/http/fault/v3/fault.proto [hunk 2]
```diff
@@ -41,6 +42,9 @@ message FaultAbort {
     // HTTP status code to use to abort the HTTP request.
     uint32 http_status = 2 [(validate.rules).uint32 = {lt: 600 gte: 200}];
 
+    // gRPC status code to use to abort the gRPC request.
+    uint32 grpc_status = 5 [(validate.rules).uint32 = {lte: 16}];
```
Comment (htuch) [inline api/envoy/extensions/filters/http/fault/v3/fault.proto:46]: @tonya11en do you recall where we ended up with gRPC as int vs. enum?

Comment (tonya11en) [reply api/envoy/extensions/filters/http/fault/v3/fault.proto:46]: We landed on using integers, so I scrapped the enum.

**Comment by vijayrajput1** [comment]
Messed my pull request commit history by doing merge and then followed my rebasing (to fix missing DCO on one of my commit 🤦‍♂️). I'll close this PR and start a new one.

**Pull request closed without merging**